File: src/routes/text.ts
import { Hono } from "hono";

const text = new Hono();

// GET /text/reverse?text=string
text.get("/reverse", (c) => {
    const text = c.req.query("text");

    if (!text) {
        return c.json({ error: { message: "No text provided", code: 400 } }, 400);
    }

    // url decode the text
    const decoded = decodeURIComponent(text);

    return c.json({ reversed: decoded.split("").reverse().join(""), original: decoded });
});

// GET /text/length?text=string
text.get("/length", (c) => {
    const text = c.req.query("text");

    if (!text) {
        return c.json({ error: { message: "No text provided", code: 400 } }, 400);
    }

    // url decode the text
    const decoded = decodeURIComponent(text);

    return c.json({
        length: {
            characters: decoded.length,
            words: decoded.split(" ").length
        },
        text: decoded
    });
});

// GET /text/uppercase?text=string
text.get("/uppercase", (c) => {
    const text = c.req.query("text");

    if (!text) {
        return c.json({ error: { message: "No text provided", code: 400 } }, 400);
    }

    // url decode the text
    const decoded = decodeURIComponent(text);

    return c.json({ uppercase: decoded.toUpperCase(), original: decoded });
});

// GET /text/lowercase?text=string
text.get("/lowercase", (c) => {
    const text = c.req.query("text");

    if (!text) {
        return c.json({ error: { message: "No text provided", code: 400 } }, 400);
    }

    // url decode the text
    const decoded = decodeURIComponent(text);

    return c.json({ lowercase: decoded.toLowerCase(), original: decoded });
});

// GET /text/replace?text=string&search=string&replace=string
text.get("/replace", (c) => {
    const text = c.req.query("text");
    const search = c.req.query("search");
    const replace = c.req.query("replace");

    if (!text || !search || !replace) {
        return c.json({ error: { message: "Missing text, search, or replace", code: 400 } }, 400);
    }

    // url decode the text
    const decoded = decodeURIComponent(text);

    return c.json({ replaced: decoded.split(search).join(replace), original: decoded, search, replace });
});

export default text;

File: routes/text.go
package routes

import (
	"strconv"
	"strings"

	"github.com/gofiber/fiber/v2"
)

// Reverse a string (text) query param
func Reverse(c *fiber.Ctx) error {
	text := c.Query("text")

	reversed := ""

	for i := len(text) - 1; i >= 0; i-- {
		reversed += string(text[i])
	}

	return c.JSON(fiber.Map{
		"reversed": reversed,
		"original": text,
	})
}

// Calculate the length of a string (text) query param
func Length(c *fiber.Ctx) error {
	text := c.Query("text")

	words := len(strings.Split(text, " "))
	characters := len(text)

	return c.JSON(fiber.Map{
		"length": fiber.Map{
			"words":      words,
			"characters": characters,
		},
		"text": text,
	})
}

// Return upercase of a string (text) query param
func Uppercase(c *fiber.Ctx) error {
	text := c.Query("text")

	uppercase := strings.ToUpper(text)

	return c.JSON(fiber.Map{
		"uppercase": uppercase,
		"original":  text,
	})
}

// Return lowercase of a string (text) query param
func Lowercase(c *fiber.Ctx) error {
	text := c.Query("text")

	lowercase := strings.ToLower(text)

	return c.JSON(fiber.Map{
		"lowercase": lowercase,
		"original":  text,
	})
}

// Replace all occurences of ?search with ?replace in ?text
func Replace(c *fiber.Ctx) error {
	text := c.Query("text")
	search := c.Query("search")
	replace := c.Query("replace")

	replaced := strings.Replace(text, search, replace, -1)

	return c.JSON(fiber.Map{
		"replaced": replaced,
		"original": text,
	})
}

// Generate a given number of lorem ipsum words
func LoremIpsum(c *fiber.Ctx) error {
	length, err := strconv.Atoi(c.Query("length"))
	if err != nil {
		return c.SendStatus(400)
	}

	lorem := "Lorem ipsum dolor sit amet consectetur adipiscing elit sed do eiusmod tempor incididunt ut labore et dolore magna aliqua Ut enim ad minim veniam quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur Excepteur sint occaecat cupidatat non proident sunt in culpa qui officia deserunt mollit anim id est laborum"
	loremWords := strings.Split(lorem, " ")

	result := ""
	for i := 0; i < length; i++ {
		result += loremWords[i%len(loremWords)] + " "
	}

	return c.JSON(fiber.Map{
		"lorem":  result,
		"length": length,
	})
}
